runtime/include/cloe/data_broker.hpp

```diff
@@ -0,0 +1,1099 @@
+    return const_cast<signal_container_t&>(const_cast<const DataBroker*>(this)->signals());
+  }
+
+  /**
+   * Subscribe to value-changed events.
+   *
+   * \tparam T Type of the signal
+   * \param name Name of the signal
+   * \param callback event-function which will be called when the value changed
+   */
+  template <typename T>
+  void subscribe(std::string_view name, databroker::on_value_changed_callback_t<T> callback) {
+    assert_static_type<T>();
+    using compatible_type = databroker::compatible_base_t<T>;
+
+    signal(name)->template subscribe<compatible_type>(std::move(callback));
+  }
+
+  /**
+   * Set the value of a signal.
+   *
+   * \tparam T Type of the signal
+   * \param name Name of the signal
+   * \param value Value to be assigned to the signal
+   */
+  template <typename T>
+  void set_value(std::string_view name, databroker::signal_type_cref_t<T> value) {
+    assert_static_type<T>();
+    using compatible_type = databroker::compatible_base_t<T>;
+
+    signal(name)->set_value<compatible_type>(value);
+  }
+
+  /**
+   * Return the value of a signal.
+   *
+   * \tparam T Type of the signal
+   * \param name Name of the signal
+   * \return Pointer to the value of the signal
+   * \note databroker::compatible_base_t<T> == T, if the function compiles
+   */
+  template <typename T>
+  databroker::signal_type_cref_t<T> value(std::string_view name) {
+    assert_static_type<T>();
+    using compatible_type = databroker::compatible_base_t<T>;
+
+    return signal(name)->value<compatible_type>();
+  }
+
+  /**
+   * Return the getter-function of a signal.
+   *
+   * \tparam T Type of the signal
+   * \param name Name of the signal
+   * \return getter-function of the signal
+   */
+  template <typename T>
+  const Signal::typed_get_value_function_t<T>& getter(std::string_view name) {
+    assert_static_type<T>();
+    using compatible_type = databroker::compatible_base_t<T>;
+
+    const Signal::typed_get_value_function_t<T>* getter_fn =
+        signal(name)->getter<compatible_type>();
+    if (!getter_fn) {
+      throw std::logic_error(fmt::format("getter for signal not provided: {}", name));
+    }
+    return *getter_fn;
+  }
+
+  /**
+   * Return the setter-function of a signal.
+   *
+   * \tparam T Type of the signal
+   * \param name Name of the signal
+   * \return const Signal::typed_set_value_function_t<T>&, setter-function of the signal
+   */
+  template <typename T>
+  const Signal::typed_set_value_function_t<T>& setter(std::string_view name) {
+    assert_static_type<T>();
+    using compatible_type = databroker::compatible_base_t<T>;
+
+    const Signal::typed_set_value_function_t<T>* setter_fn =
+        signal(name)->setter<compatible_type>();
+    if (!setter_fn) {
+      throw std::logic_error(fmt::format("setter for signal not provided: {}", name));
+    }
+    return *setter_fn;
+  }
+};
+
+/**
+ * Return a reference to the central instance of the databroker
+ */
+extern DataBroker& db();
```

Comment:
I don't like the idea of a global DataBroker instance. Global state is an established anti-pattern. Every agent/model interacting with a Cloe simulation has an `enroll(Registrar&)` method. That is where I would make the `DataBroker` available.

Is such a late interaction with the data broker database possible? Otherwise we can consider providing it at construction already. This will require breaking changes, but we have those anyway.

Reply:
Here is some discussion on global state and also what alternatives exist: https://softwareengineering.stackexchange.com/questions/148108/why-is-global-state-so-evil

I am advocating for dependency injection as a solution.

Reply:
Sounds good.

The basic design goal was to grant hazzle free access to the one relevant instance. Another implementation could be e.g. in certain objects which are anyhow widely available.

Reply:
Should be hassle-free.